pika/adapters/base_connection.py

```diff
@@ -239,7 +241,10 @@ def _get_error_code(self, error_value):
         return None
 
     def _flush_outbound(self):
-        """Call the state manager who will figure out that we need to write."""
+        """write early, if the socket will take the data why not get it out
+        there asap.
+        """
+        self._handle_write()
```

Comment:
Yup sure, I'm not casting any aspersions in pika's `_handle_read` overhead.

Perhaps my original comment was too flippant. I am actually _extremely_ interested in the performance of pika, but more than that, I'm interested in getting behaviour that is reliable, consistent and what a user might reasonably expect. I think this PR gets us a step closer to that goal, AND it is actually significantly more efficient than the existing code.


Let's keep in mind that we're only talking about a situation here where the publisher is consistently generating messages faster than the network can send them, in this situation we would actually be pushing back using the backpressure detection, if it worked (I have a commit for that too).  In such a situation, taken to extreme, we're talking about swapping and potentially memory exhaustion, IMO a small additional overhead (1.4us* on my laptop) is probably not the first thing to worry about.

Furthermore, scenarios such as the following continue to break down...

Imagine you have an application that is reading a large amount of data from a non rabbit source (say a large XML document retrieved over http, data from some sort of DB or even generated internally), as it processes that data it publishes a large number of messages. In this situation the code might well not drop into the ioloop until it's completed processing of all the data. On pika/master it would never attempt to send a single message until processing of the full document was complete whereas this PR would keep putting messages in the socket buffer whenever space was available. With your suggestion we would fill the socket buffer once and then never send another message until we finished processing the document.

My aim with this PR is to get the data on the wire ASAP to minimise latency and maximise opportunities for parallel-isation.

*1.4us is the overhead on my laptop of attempting a 1byte write on a blocked socket in python, including catching and handling the generated exception.
0.5us is the overhead of an equivalent write on an unblocked socket.
4.3us is the overhead of dropping into the tornado ioloop and getting triggered when the socket becomes writeable as the existing code does, individually for EVERY frame sent.